### PR TITLE
fix(inductor): select SDPA tiling from LX footprint

### LINKED ISSUE #4339: Encoder fused SDPA 3.6× slower at L=512 after #4259 (e02b78ba → 5e715ea)
## Summary

`torch.compile` of fused (unmasked) `F.scaled_dot_product_attention` got ~3.6× slower between:

- **fast:** `e02b78ba35f9a1a69a458c3149e9c01d9f4fa6a8`
- **slow:** `5e715ea70c4aa9e76e45b7fbae6b92a6d7658764` (#4259 SDPA rewrite, 34 commits later)

Isolated to the **fused** kernel. Packed QK and P·V on the same shape/dtype are unchanged ( on both SHAs). Repro is a single `torch.compile` of permute + non-causal `F.sdpa` — no spyre-inference.

## When it shows up

Compiling encoder self-attention (granite-embedding-125m) on Spyre:

- `B=1`, `H=12`, `L=512`, `D=64`, fp16
- non-causal `F.sdpa`, **no `attn_mask`**
- layout is `[L, H, D]` → permute to `[1, H, L, D]` then fused SDPA (the B=1 fused path used when the prompt fills the L=512 bucket)

The slowdown appears as soon as that fused kernel is compiled and run on `5e715ea`. The same compile on `e02b78ba` is fast. Packed (masked) QK / P·V at the same shape are not affected.

## Isolated microbench

Shape `[512, 12, 64]` fp16, Spyre, `torch.compile(dynamic=False)`:

#4259 dropped the old `work_div` tiling. At L=512 the new decomp uses one 512-wide Q tile. Forcing tile=64 on the **new** decomp unrolls 8×8 online-softmax without the old `work_div` path and is worse, not a workaround.

## Repro

Run the same script against the two torch-spyre SHAs above. Wipe inductor cache when switching version or tile size.

```bash
rm -rf /tmp/torchinductor_* ~/.cache/torchinductor
python repro_fused_sdpa.py
```

Optional, **on `5e715ea` only** — shows shrinking the tile cap is not a fix:

```bash
rm -rf /tmp/torchinductor_* ~/.cache/torchinductor
python repro_fused_sdpa.py --sdpa-tile-size 64
```

## Script (`repro_fused_sdpa.py`)

```python
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import os
import statistics
import time

os.environ.setdefault("TORCH_DEVICE_BACKEND_AUTOLOAD", "0")
os.environ.setdefault("RANK", "0")
os.environ.setdefault("LOCAL_RANK", "0")
os.environ.setdefault("WORLD_SIZE", "1")
os.environ.setdefault("LOCAL_WORLD_SIZE", "1")

import torch
import torch.nn.functional as F


def fused_sdpa(q, k, v, scale):
    q = q.unsqueeze(0).transpose(1, 2)
    k = k.unsqueeze(0).transpose(1, 2)
    v = v.unsqueeze(0).transpose(1, 2)
    out = F.scaled_dot_product_attention(q, k, v, is_causal=False, scale=scale)
    t, h, d = q.shape[2], q.shape[1], q.shape[3]
    return out.transpose(1, 2).reshape(t, h, d)


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("--seq", type=int, default=512)
    p.add_argument("--heads", type=int, default=12)
    p.add_argument("--dim", type=int, default=64)
    p.add_argument("--warmup", type=int, default=3)
    p.add_argument("--iters", type=int, default=20)
    p.add_argument("--sdpa-tile-size", type=int, default=None)
    args = p.parse_args()

    import torch_spyre

    torch_spyre._autoload()

    tile = None
    try:
        import torch_spyre._inductor.decompositions as decomp

        tile = getattr(decomp, "_SDPA_MAX_SEQUENCE_TILE_SIZE", None)
        if args.sdpa_tile_size is not None and tile is not None:
            decomp._SDPA_MAX_SEQUENCE_TILE_SIZE = args.sdpa_tile_size
            tile = args.sdpa_tile_size
    except Exception:
        pass

    device = torch.device("spyre")
    dtype = torch.float16
    scale = args.dim ** -0.5
    gen = torch.Generator().manual_seed(0)
    q = torch.randn(args.seq, args.heads, args.dim, dtype=dtype, generator=gen).to(device)
    k = torch.randn(args.seq, args.heads, args.dim, dtype=dtype, generator=gen).to(device)
    v = torch.randn(args.seq, args.heads, args.dim, dtype=dtype, generator=gen).to(device)

    fn = torch.compile(fused_sdpa, dynamic=False)
    print(f"tile={tile}  [{args.seq},{args.heads},{args.dim}]")
    t0 = time.perf_counter()
    fn(q, k, v, scale)
    torch.spyre.synchronize()
    print(f"compile+first {time.perf_counter() - t0:.1f}s")

    for _ in range(args.warmup):
        fn(q, k, v, scale)
        torch.spyre.synchronize()
    times = []
    for _ in range(args.iters):
        torch.spyre.synchronize()
        t0 = time.perf_counter()
        fn(q, k, v, scale)
        torch.spyre.synchronize()
        times.append((time.perf_counter() - t0) * 1000)
    print(f"fused ms_mean={statistics.mean(times):.3f}  min={min(times):.3f}")


if __name__ == "__main__":
    main()
```

### PR DESCRIPTION
**Pull request by ani300**
## Summary

- add a pure SDPA tiling selector driven by MHA/GQA shape, exact 32-core work division, and per-core LX footprint
- use the one-block, non-coarse-head-tiled path for the issue #4339 shape while preserving the existing post-#4259 strategy for GQA and long contexts
- log the selected strategy, rejection reason, work division, and estimated per-core score footprint
- add CPU-only policy coverage for #4339, Granite GQA, 8K/32K contexts, divisibility, LX pressure, and alternate core counts

Fixes #4339.

## Validation

- `python -m pytest -q tests/inductor/test_sdpa_tiling.py tests/inductor/test_building_blocks.py::TestBuildingBlocks::test_sdpa_head_tiles_limit_heads_per_tile`
- `python -m pytest -q tests/inductor/test_building_blocks.py::TestBuildingBlocks::test_causal_sdpa_unpadded_kv_no_inf tests/inductor/test_building_blocks.py::TestBuildingBlocks::test_granite_gqa_prefill_with_finite_broadcast_mask`
- exact #4339 repro: max abs error 0.003662, mean abs error 0.000182, 0.380 ms mean / 0.353 ms minimum wall time, 24 SDSCs, no `LoopSpec`, 164 KiB binary

### CODE REVIEW

**Comment by ani300** [comment]
Decode follow-up is now pushed in 786842cf.

I swept the production SDPA decomposition with the real Granite 3.3 8B and Gemma 4 12B/26B tensor layouts at Lq=1 and Lq=512, for Lkv=512, 1024, and 8192. Configurations were ranked by average Spyre kernel time; correctness and no-recompile checks passed unless noted.

Selected decode policies and 8K comparison against the prior K512/no-work-division policy:

| Geometry | New policy | Old 8K | New production 8K | Change |
|---|---|---:|---:|---:|
| Granite 32H/8KV/D128 | full H, K4096 cap | 6.348 ms | 5.935 ms | -6.5% |
| Gemma local 16H/8KV/D256 | full H, K2048 cap | 7.735 ms | 5.131 ms | -33.7% |
| Gemma 12B global 16H/1KV/D512 | H16 work split, K512 cap | 11.629 ms | 1.311 ms | -88.7% |
| Gemma 26B global 16H/2KV/D512 | H8 work split, K256 cap | 17.148 ms | 6.344 ms | -63.0% |

The largest fitting tile was not generally fastest. Examples: Granite decode at 8K favored K4096 over K8192 (5.895 vs 9.304 ms in the forced sweep); Gemma local favored K2048 over K4096/K8192 (5.156 vs 7.350/10.670 ms); Gemma global hits a severe cliff above its selected caps.

Prefill selections remain unchanged because the measured winners were stable at all three context lengths:

| Geometry | Policy | 512 | 1024 | 8192 |
|---|---|---:|---:|---:|
| Granite | K512, H4 x Q8 | 0.599 | 1.116 | 7.182 ms |
| Gemma local | K256, H2 x Q16 | 0.781 | 1.368 | 9.604 ms |
| Gemma 12B global | K512, H8 x Q4 | 1.048 | 1.677 | 10.308 ms |
| Gemma 26B global | K512, H8 x Q4 | 1.178 | 2.335 | 15.749 ms |

Lkv work-division values were timing-equivalent in prefill; H/Q work division is what mattered. All chosen per-iteration live-set estimates remain below the 1,625,344-byte frontend LX budget.

Model-level second-run results with 512-token prompt and 5 generated tokens:

- Granite 3.3 8B: TTFT 381.9 ms, decode 162.4 ms/token, deterministic PASS. Earlier decode was about 175 ms/token.
- Gemma 4 12B: TTFT 1273.2 ms, decode 223.9 ms/token, E2E 2.214 s, deterministic PASS. Previous values were 1275.0 ms, 245.4 ms/token, and 2.303 s.

Validation: 13 selector tests plus 20 subtests pass; all four production-selector 8K kernels pass correctness, apply their requested hints, and do not recompile. The CI config integrity check also passes.

**Review by dgrove-oss — APPROVED** [review]
Summary

Fixes issue #4339 by replacing an ad-hoc SDPA tiling heuristic with a cost-model-driven selection (_select_sdpa_tiling()) that picks among work_divided, work_divided_tiled, coarse_tiled, and three decode-specific strategies (decode, decode_tiled, decode_work_divided, decode_work_divided_tiled) based on estimated per-core LX footprint versus the real LX budget, plus a calibrated lookup table for single-query decode geometries.

Review

BLOCKERs

(none)

SUGGESTIONs

- Consider a brief comment at _SDPA_DECODE_POLICIES' definition site cross-referencing how the calibrated entries were derived (empirical tuning vs. formula), so future maintainers know whether adding a new (num_heads, num_kvheads, head_dim) entry requires re-running a calibration sweep or can be computed. (Not blocking — the intent is clear from context, but the provenance of magic-looking tuples is easy to lose over time.)

QUESTIONs

- _sdpa_estimated_live_bytes_per_core() uses fixed allowances (_SDPA_LIVE_SCORE_BUFFER_ALLOWANCE=2, _SDPA_LIVE_QUERY_BUFFER_ALLOWANCE=2) as safety margins against the LX budget. Are these derived from a specific worst-case double-buffering scenario, or empirically chosen headroom? Worth confirming they'll still be conservative enough if head_dim or KV-block sizing assumptions change later.
- _SDPA_MHA_MAX_HEAD_WORK_DIVISION=4 vs. _SDPA_GQA_MAX_HEAD_WORK_DIVISION=8 — is the MHA cap intentionally half the GQA cap because of a specific core-count/head-count interaction, or is this just what was validated so far? Not blocking, just want to confirm it's not an arbitrary placeholder.

Verification performed

- Independently re-derived _lx_planning_size()'s real default value from its constituent constants (_LX_TRACKER_CAPACITY_BYTES = (2<<20) - (64<<10) = 2,031,616, DXP_LX_FRAC_AVAIL=0.2, rounded up to 128-byte alignment) and confirmed it matches exactly the test suite's hardcoded _LX_BUDGET = 1_625_344 constant in tests/inductor/test_sdpa_tiling.py — the test suite is grounded in the actual runtime default, not an arbitrary number.
- Hand-traced the numeric logic of _select_sdpa_tiling(), _sdpa_full_core_work_division(), _sdpa_kv_work_division(), and _sdpa_estimated_live_bytes_per_core() against several of the new test cases (test_issue_4339_uses_one_block_and_work_division, test_granite_gqa_uses_calibrated_work_divided_tile, test_decode_uses_geometry_calibrated_policy, test_lx_budget_reduces_kv_block_until_live_values_fit) and confirmed the code produces the asserted tiling choices.
- Confirmed license header present and correct on the new test file (tests/inductor/test_sdpa_tiling.py, 283 lines).
- Confirmed spyre__sdpa_overrideable()'s integration point correctly uses contextlib.nullcontext() when tiling.work_div is None, avoiding an unnecessary hint scope when no explicit work-division is chosen.
- This is compiled-path only (torch_spyre/_inductor/decompositions.py), consistent with test coverage added only under tests/inductor/.
- DCO and pre-commit compliance not flagged as issues based on gh pr checks/gh pr view output.

Verdict

APPROVE (pending final CI completion — no failures observed as of last check; will confirm unconditional once green)
